[AMDGPU] Enable reordering of VMEM loads during clustering

perlfu opened the pull request:
Add fine grain control over ReorderWhileClustering by adding
canReorderClusterMemOps query to TargetInstrInfo.
Implement this to true for RISC to maintain current behaviour.

On AMDGPU enable ReorderWhileClustering for loads and implement
canReorderClusterMemOps to reject reordering for operations other
than VMEM.
The intention of doing this is to allow some additional overlap
of computation with memory loads with large clauses.
Loads will be issued in an order closer to their usage more often,
producing more incremental s_waitcnt values.
On average this yields a very small reduction in VGPR pressure,
although edge cases may see increased pressure.
Reordering SMEM/LDS access is not beneficial as these must always
be waitcnt 0.

For the benefit of future tuning add support for function metadata
"amdgpu-reorder-loads-while-clustering" (set to either 0 or 1) to
disable/enable reordering behaviour per function.

arsenm commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -581,6 +581,23 @@ bool SIInstrInfo::shouldClusterMemOps(ArrayRef<const MachineOperand *> BaseOps1,
+      !BaseOps1.empty() ? BaseOps1.front()->getParent() : nullptr;
+  const MachineInstr *SecondLdSt =
+      !BaseOps2.empty() ? BaseOps2.front()->getParent() : nullptr;
+  ;
```
Stray ;

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -483,12 +488,20 @@ static ScheduleDAGInstrs *createSIMachineScheduler(MachineSchedContext *C) {
+static bool getReorderWhileLoadClustering(const MachineFunction *MF) {
+  if (!ReorderWhileLoadClustering)
+    return false;
+  return !!MF->getFunction().getFnAttributeAsParsedInteger(
+      "amdgpu-reorder-loads-while-clustering", 1);
```
If you really want to add another one of these explicit boolean attributes, should use getValueAsBool and "true" "false"

Do we really need both mechanisms here? At most I would think the debug cl::opt may be useful for a short period

arsenm commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -581,6 +581,23 @@ bool SIInstrInfo::shouldClusterMemOps(ArrayRef<const MachineOperand *> BaseOps1,
+  const MachineInstr *SecondLdSt =
+      !BaseOps2.empty() ? BaseOps2.front()->getParent() : nullptr;
+  ;
+
```
Comment this

arsenm commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -581,6 +581,23 @@ bool SIInstrInfo::shouldClusterMemOps(ArrayRef<const MachineOperand *> BaseOps1,
+    return false;
+  if (!SecondLdSt || !isVMEM(*SecondLdSt))
+    return false;
+
```
Does the calling code take care of correctness / aliasing issues?

perlfu committed:
[AMDGPU] Enable reordering of VMEM loads during clustering

The intention of doing this is to allow some additional overlap
of computation with memory loads, as loads will be issued in
an order closer to their usage, more incremental s_waitcnt can
be introduced.
On average this yields a very small reduction in VGPR pressure,
although edge cases may see increased pressure.

For the benefit of future tuning add support for function metadata
"amdgpu-reorder--while-clustering" to disable/enable reordering
behaviour per function.

arsenm reviewed (commented):
Can you do this with areMemAccessesTriviallyDisjoint? Also due for rebase

perlfu committed:
- Test changes

arsenm commented on llvm/lib/CodeGen/MachineScheduler.cpp:
```diff
@@ -2120,8 +2120,11 @@ void BaseMemOpClusterMutation::clusterNeighboringMemOps(
+    bool CanReorder = false;
+    if (ReorderWhileClustering)
+      CanReorder = TII->areMemAccessesTriviallyDisjoint(*SUa->getInstr(),
+                                                        *SUb->getInstr());
```
```suggestion
    bool CanReorder = ReorderWhileClustering &&
      TII->areMemAccessesTriviallyDisjoint(*SUa->getInstr(),
                                                        *SUb->getInstr());
```

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -703,12 +708,21 @@ static ScheduleDAGInstrs *createSIMachineScheduler(MachineSchedContext *C) {
+  if (!ReorderWhileClustering)
+    return false;
+  Attribute FnAttr =
+      MF->getFunction().getFnAttribute("amdgpu-reorder-while-clustering");
```
I wouldn't add the attribute without a specific need

perlfu committed:
- Reviewer comments

perlfu commented:
- Rebase
- Address reviewer feedback
